src/algorithms/rsa.rs

```diff
@@ -151,38 +174,40 @@ pub fn rsa_decrypt_and_check<R: CryptoRngCore + ?Sized>(
 }
 
 /// Returns the blinded c, along with the unblinding factor.
-fn blind<R: CryptoRngCore, K: PublicKeyParts>(
+fn blind<R: CryptoRngCore, K: PublicKeyPartsNew>(
     rng: &mut R,
     key: &K,
-    c: &BigUint,
-) -> (BigUint, BigUint) {
+    c: &BoxedUint,
+    n_params: &BoxedResidueParams,
+) -> (BoxedUint, BoxedUint) {
     // Blinding involves multiplying c by r^e.
     // Then the decryption operation performs (m^e * r^e)^d mod n
     // which equals mr mod n. The factor of r can then be removed
     // by multiplying by the multiplicative inverse of r.
 
-    let mut r: BigUint;
-    let mut ir: Option<BigInt>;
+    let mut r: BoxedUint;
+    let mut ir: CtOption<BoxedUint>;
     let unblinder;
     loop {
-        r = rng.gen_biguint_below(key.n());
-        if r.is_zero() {
-            r = BigUint::one();
+        // TODO: use constant time gen
+        r = BoxedUint::random_mod(rng, key.n());
+        // TODO: correct mapping
+        if r.is_zero().into() {
+            r = BoxedUint::one();
         }
-        ir = r.clone().mod_inverse(key.n());
-        if let Some(ir) = ir {
-            if let Some(ub) = ir.into_biguint() {
-                unblinder = ub;
-                break;
-            }
+        ir = r.inv_mod(key.n());
+
+        // TODO: constant time?
```

Comment:
Ordinarily I'd suggest `CtOption::map` here but it doesn't work on `BoxedUint` due to a `Copy` bound required for `ConditionallySelectable`: https://github.com/dalek-cryptography/subtle/issues/94#issuecomment-1829905653

We might need to add a little library of helper functions to work around it, however it's complicated by the need to respect the precision of the original value. I could maybe PR something to `crypto-bigint` as we might end up needing it there.

Reply:
I added some constant time combinators for `BoxedUint`, but I now realize that they wouldn't help in this case:

https://github.com/RustCrypto/crypto-bigint/pull/352

However, they're now available if you do end up needing them.